File: dataforge-core/src/main/java/hep/dataforge/io/markup/Markup.java
package hep.dataforge.io.markup;

import hep.dataforge.description.Described;
import hep.dataforge.description.NodeDef;
import hep.dataforge.description.ValueDef;
import hep.dataforge.meta.Configuration;
import hep.dataforge.meta.Laminate;
import hep.dataforge.meta.Meta;
import hep.dataforge.meta.SimpleConfigurable;
import hep.dataforge.utils.Optionals;
import hep.dataforge.values.Value;
import hep.dataforge.values.ValueProvider;
import org.jetbrains.annotations.Nullable;

import java.util.Optional;
import java.util.function.Function;
import java.util.stream.Stream;

/**
 * Basic markup element wrapper. A markup element must have a type designation, fields specific for this markup element (like 'text' for text markup).
 * The structure of intrinsic meta could be as complex as needed as long as its upper layer nodes do not have name conflict with standard markup nodes like 'style'
 * 'style' element contains optional style information.
 * Created by darksnake on 30-Dec-16.
 */
@ValueDef(name = "type", info = "The type of this block")
@NodeDef(name = "meta", info = "Meta specific for this element")
@NodeDef(name = "style", info = "Style override")
public class Markup extends SimpleConfigurable implements Described, ValueProvider {
    /**
     * A generic container type.
     */
    public static final String MARKUP_GROUP_TYPE = "group";
    public static final String MARKUP_STYLE_NODE = "style";
    public static final String MARKUP_CONTENT_NODE = "c";
    public static final String MARKUP_TYPE_KEY = "type";

    public Markup(Configuration c) {
        super(c);
    }

    public Markup(Meta meta) {
        super(new Configuration(meta));
    }

    /**
     * Get type of this block. If type is not defined use externally inferred type.
     *
     * @return
     */
    public String getType(Function<Markup, String> infer) {
        return meta().getString(MARKUP_TYPE_KEY, () -> infer.apply(this));
    }

    /**
     * Get type of this block. If type is not defined, use group type.
     *
     * @return
     */
    public String getType() {
        return meta().getString(MARKUP_TYPE_KEY, MARKUP_GROUP_TYPE);
    }

    /**
     * Get the parent element for this one. If null, then this is a root element
     *
     * @return
     */
    @Nullable
    public Markup getParent() {
        //TODO add caching to avoid reconstruction of the tree each time this method is called
        if (getConfig().getParent() == null) {
            return null;
        } else {
            return new Markup(getConfig().getParent());
        }
    }

    /**
     * Style ignores values outside {@code style} node
     * @return
     */
    public Laminate getStyle() {
        Laminate laminate = new Laminate(meta()
                .getMetaOrEmpty(MARKUP_STYLE_NODE))
                .withDescriptor(getDescriptor());

        Markup parent = getParent();

        if (parent != null) {
             laminate = laminate.withLayer(parent.getStyle());
        }
        return laminate;
    }

    @Override
    protected void applyConfig(Meta config) {
        //do nothing
    }

    @Override
    public Optional<Value> optValue(String path) {
        return Optionals.either(meta().optValue(path)).or(()->getStyle().optValue(path)).opt();
    }

    /**
     * Stream of child nodes of this node in case it serves as a group
     *
     * @return
     */
    public Stream<Markup> getContent() {
        return getConfig().getMetaList(MARKUP_CONTENT_NODE).stream().map(it -> new Markup(it));
    }

}

File: dataforge-core/src/main/kotlin/hep/dataforge/io/markup/Markup.kt
package hep.dataforge.io.markup

import hep.dataforge.description.NodeDef
import hep.dataforge.description.NodeDefs
import hep.dataforge.description.ValueDef
import hep.dataforge.io.markup.Markup.Companion.MARKUP_CONTENT_NODE
import hep.dataforge.io.markup.Markup.Companion.MARKUP_STYLE_NODE
import hep.dataforge.io.markup.Markup.Companion.MARKUP_TYPE_KEY
import hep.dataforge.kodex.node
import hep.dataforge.meta.Laminate
import hep.dataforge.meta.Meta
import hep.dataforge.meta.MetaMorph
import hep.dataforge.meta.MetaNode.DEFAULT_META_NAME
import hep.dataforge.meta.MorphProvider
import hep.dataforge.values.Value
import hep.dataforge.values.ValueProvider
import java.util.*


/**
 * Basic markup element wrapper. A markup element must have a type designation, fields specific for this markup element (like 'text' for text markup).
 * The structure of intrinsic meta could be as complex as needed as long as its upper layer nodes do not have name conflict with standard markup nodes like 'style'
 * 'style' element contains optional style information.
 * Created by darksnake on 30-Dec-16.
 */

@ValueDef(name = MARKUP_TYPE_KEY, def = Markup.MARKUP_GROUP_TYPE, info = "The type of this block")
@NodeDefs(
        NodeDef(name = DEFAULT_META_NAME, info = "Meta specific for this element"),
        NodeDef(name = MARKUP_STYLE_NODE, info = "Style override")
)
interface Markup : MetaMorph, ValueProvider {

    val parent: Markup?
    val type: String
    val content: List<Markup>
    /**
     * Private style of this markup
     */
    val style: Meta

    /**
     * Set of styles including all ancestors
     * //TODO better name
     */
    val styleStack: Laminate
        get() {
            return parent?.styleStack?.withFirstLayer(style) ?: Laminate(style)
        }

    companion object : MorphProvider<Markup> {
        /**
         * A generic container type.
         */
        const val MARKUP_GROUP_TYPE = "group"
        const val MARKUP_STYLE_NODE = "style"
        const val MARKUP_CONTENT_NODE = "content"
        const val MARKUP_TYPE_KEY = "type"

        const val TEXT_TYPE = "text"
        const val HEADER_TYPE = "head"
        const val LIST_TYPE = "list"
        const val TABLE_TYPE = "table"
        const val ROW_TYPE = "tr"

        override fun morph(meta: Meta): Markup {
            return morph(meta, null)
        }

        fun morph(meta: Meta, parent: Markup? = null): Markup {
            return GenericMarkup(meta, parent)
        }

        fun inferType(element: Markup): String {
            return if (element.hasValue("text")) {
                TEXT_TYPE
            } else {
                when (element.parent?.type) {
                    TABLE_TYPE -> ROW_TYPE
                    else -> throw RuntimeException("Can't infer markup element type")
                }
            }
        }
    }
}

class GenericMarkup(val meta: Meta, override val parent: Markup? = null) : Markup {

    override fun optValue(path: String): Optional<Value> {
        return meta.optValue(path)
    }

    override val type = meta.getString("type") { Markup.inferType(this) }

    override val content: List<Markup>
        get() = meta.getMetaList(MARKUP_CONTENT_NODE).map { Markup.morph(it, this) }

    override val style by meta.node()

    override fun toMeta(): Meta = meta
}

